### ISSUE TITLE
Unpredictable Crash on App Start Due to Test Observer Registration Outside Main Thread

### ISSUE THREAD

**Issue by Atimca** [pointfreeco/swift-dependencies#100]
### Description

The **TCA based** application is unpredictably crashing upon start during the configuration load. The crash occurs on the line `let config = try? await remoteConfigService.fetchRemoteConfig()` from the AppConfig module, producing the error message: '**Test observers can only be registered and unregistered on the main thread.**' 

On further investigation, it was found that the crash likely happens during the testing environment check in `DependencyValues.swift` of the Dependencies framework, specifically at `else if _XCTIsTesting`. The crash seems to continue at the initialization of `XCTCurrentTestCase` where the test observers are launched.

**Additional Notes:**
- The bug does not occur all the time but is more prevalent on slower machines compared to more powerful ones.
- It is suggested that the issue lies within the `XCTCurrentTestCase` where test observers are being initialized.

**Temporary workaround that helped:**
```swift
        case .loadConfig:
            return .run { @MainActor send in
                let config = try? await remoteConfigService.fetchRemoteConfig()
                send(.configLoaded(config))
            }
```

Implementation of **AppFeature**, **AppConfigFeature** and **RemoteConfigService**:

```swift
struct AppFeature: ReducerProtocol {
    
    // MARK: - State
    struct State: Equatable {
        var appConfig: AppConfigFeature.State
    }
    
    // MARK: - Action
    enum Action: Equatable {
        case appDidFinishLaunching

        case appConfig(AppConfigFeature.Action)
    }

    // MARK: - Dependencies
    @Dependency(\.datadogActivator)
    var datadogActivator
    
    // MARK: - Reducer
    var body: some ReducerProtocol<State, Action> {
        Scope(state: \.appConfig, action: /Action.appConfig, child: AppConfigFeature.init)
        Reduce { state, action in
            switch action {
            case .appDidFinishLaunching:
                return .run { send in await send(Action.appConfig(.loadConfig)) }
            case .appConfig(.configLoaded):
                guard let info = state.appConfig.config?.dataDog else {
                    assertionFailure()
                    return .none
                }
                return .run { _ in await datadogActivator.activate(info: info) }
            case .appConfig:
                return .none
            }
        }
    }
}
```

```swift
public struct AppConfigFeature: ReducerProtocol, Sendable {

    // MARK: - State
    public struct State: Equatable {
        public var config: RemoteConfigValue?
    }

    // MARK: - Action
    public enum Action: Equatable, Sendable {
        case configLoaded(RemoteConfigValue?)
        case loadConfig
    }

    // MARK: - Dependencies
    @Dependency(\.remoteConfigService)
    var remoteConfigService

    // MARK: - Reducer
    public func reduce(into state: inout State, action: Action) -> EffectTask<Action> {
        switch action {
        case .configLoaded(let config):
            state.config = config
            return .none
        case .loadConfig:
            return .run { send in
                let config = try? await remoteConfigService.fetchRemoteConfig()
                await send(.configLoaded(config))
            }
        }
    }
}
```

```swift
import Dependencies
import FirebaseRemoteConfig

struct RemoteConfigService: Sendable {
    var fetchRemoteConfig: @Sendable () async throws -> RemoteConfigValue
}

// MARK: - DependencyKey
extension RemoteConfigService: DependencyKey {
    static let liveValue: RemoteConfigService = {
        @Dependency(\.appClient.isRunningUITests)
        var isRunningUITests
        return isRunningUITests()
        ? .uiTestStub
        : .live
    }()
    static let testValue = RemoteConfigService.failing
}
extension DependencyValues {
    var remoteConfigService: RemoteConfigService {
        get { self[RemoteConfigService.self] }
        set { self[RemoteConfigService.self] = newValue }
    }
}

// MARK: - Implementations
private extension RemoteConfigService {

    static let live = Self {

        let firebaseConfig = RemoteConfig.remoteConfig()
        firebaseConfig.setDefaults(fromPlist: "RemoteConfigDefaults")

        return try await withCheckedThrowingContinuation { continuation in
            firebaseConfig.fetchAndActivate { _, error in

                if let error {
                    continuation.resume(throwing: error)
                    return
                }

                do {
                    continuation.resume(returning: try RemoteConfigValue(from: firebaseConfig))
                } catch {
                    continuation.resume(throwing: error)
                }

            }
        }
    }

    static let failing = Self(
        fetchRemoteConfig: unimplemented("\(Self.self).fetchRemoteConfig")
    )

    static let uiTestStub = Self(
        fetchRemoteConfig: { RemoteConfigValue.mock() }
    )
}
```

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [ ] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-dependencies/issues) or [discussion](https://github.com/pointfreeco/swift-dependencies/discussions).

### Expected behavior

The application should start and load the app config successfully.

### Actual behavior

The application crashes on the line `let config = try? await remoteConfigService.fetchRemoteConfig()` with the error:
```
libc++abi: terminating with uncaught exception of type NSException
*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Test observers can only be registered and unregistered on the main thread.'
terminating with uncaught exception of type NSException
CoreSimulator 857.14 - Device: iPhone 14 Pro (12DDE56D-B64A-4155-85C5-9D807F586EB2) - Runtime: iOS 16.2 (20C52) - DeviceType: iPhone 14 Pro
```

### Steps to reproduce

1. Start the application
2. `AppDelegate.didFinishLaunchingWithOptions` sends `viewStore.send(.appDidFinishLaunching)` to the AppFeature.
3. In the app reducer, the app config is loaded from a dedicated module via 
```
case .appDidFinishLaunching:
return .run { send in await send(Action.appConfig(.loadConfig)) }
```
4. Inside AppConfig module we run 
```
case .loadConfig:
return .run { send in
    let config = try? await remoteConfigService.fetchRemoteConfig()
    await send(.configLoaded(config))
}
```

### Dependencies version information

0.5.1

### Destination operating system

iOS 16.2 Simulator

### Xcode version information

14.2.0

### Swift Compiler version information

```shell
swift-driver version: 1.62.15 Apple Swift version 5.7.2 (swiftlang-5.7.2.135.5 clang-1400.0.29.51)
Target: x86_64-apple-macosx13.0
```

**Comment by stephencelis** [comment]
@Atimca I think we may need more info to diagnose and reproduce, but this sounds like it might not be a bug with the library.

One question that doesn't seem to be covered in the issue: do you encounter this crash during test runs, or are you experiencing a crash when running the app normally? I would guess you're talking about test runs since the code path causing the crash shouldn't run during the course of a normal application.

Assuming this is the case, it sounds like your tests may be running in a host application, which means the host application is exercising live dependencies while XCTest prepares its test run. Running tests in a host application is a gotcha mentioned in the documentation here:

https://pointfreeco.github.io/swift-dependencies/main/documentation/dependencies/testing#Testing-gotchas

We advise ensuring that your tests do not run your application code and attempt to interact with live dependencies. In fact, if my assumptions above are true, that means each time you run your tests you also fire off a request for the remote config, which is probably work you don't want to be doing.

If I got any of the above wrong, please let us know with more information, and ideally a fully-working repro of the issue.

The only thing actionable in the library, I think, would be for us to avoid some of these XCTest background thread crashes by synchronously sending them to the main queue instead, but ideally we would have a repro to work with to do that work.

**Comment by mbrandonw** [comment]
Hi @Atimca, do you have any update for this issue? If not we may close it soon. Thanks.

**Comment by Atimca** [comment]
Hi @stephencelis and @mbrandonw. Apologies for the delay. The issue occurred while running the app normally in debug mode. However, after considering @stephencelis's assumption that it was in the testing scenario, we decided to dig more deeply into our project. Initially, we tried to reproduce the issue in a clean project but with no success. Then, we discovered an internal framework weakly loading the XCTest framework. This appears to be the exact issue that interfered with the Dependencies (or XCTDynamicOverlay) framework upon app startup. Our investigation continues, but it seems the problem originates from our side. It's unclear if any library updates could improve this; it seems we might need to tidy up on our end.

![Screenshot 2023-06-16 at 12 23 33](https://github.com/pointfreeco/swift-dependencies/assets/13145815/ae6fdb68-19d0-4c6b-8c72-26daf0c3385a)
![Screenshot 2023-06-16 at 11 24 58](https://github.com/pointfreeco/swift-dependencies/assets/13145815/4dd82cd7-fd69-4dd0-bd61-7dadb947369d)
![Screenshot 2023-06-16 at 11 13 32](https://github.com/pointfreeco/swift-dependencies/assets/13145815/eefaa354-3cf0-44e4-92f6-9dec5b039e14)
